### PR TITLE
[DCS-151] Generate numbers

### PR DESCRIPTION
**Pull request by eruizgar**
## Description

- This feature extends Hermes with the porpouse of generate random integers and decimal numbers. The complete description of how use methods is in README.md included.

### Testing

- You could find unit test in HermesTest.scala

- [X] Unit, integration, acceptance tests as appropriate
- [X] Coverage (>90%)

### Documentation
- [X] Changelog update
- [X] Documentation entry

### Scalastyle
Result of scalastyle execution: 
Processed 16 file(s)
Found 0 errors
Found 0 warnings
Found 0 infos

### CODE REVIEW

File: src/main/scala/com/stratio/hermes/utils/Hermes.scala [hunk 1]
```diff
@@ -16,55 +16,155 @@
...
-
-import scala.util.Try
+import com.stratio.hermes.helpers.RandomHelper
+import scala.language.postfixOps
+import scala.util.{Random, Try}
 
 /**
- * Hermes util used for to generate random values.
- */
+  * Hermes util used for to generate random values.
```
Comment (albertostratio) [inline src/main/scala/com/stratio/hermes/utils/Hermes.scala]: I think we should all set-up our IDEs the same way to avoid these changes in code that we haven't modified.

File: src/main/scala/com/stratio/hermes/utils/Hermes.scala [hunk 2]
```diff
@@ -16,55 +16,155 @@
...
+      * Example: "decimal(3,2,Sign.-) -> -123.45".
+      *
+      * @param m    integer part size.
+      * @param n    decimal part size.
+      * @param sign sign positive or negative.
+      * @return a random double with defined sign.
+      */
+    def decimal(m: Int, n: Int, sign: Sign): Double = (number(m, sign).toString + "." + numberDec(n)).toDouble
+  }
+
```
Comment (albertostratio) [inline src/main/scala/com/stratio/hermes/utils/Hermes.scala]: Could you please get rid of this blank line

File: src/main/scala/com/stratio/hermes/utils/Hermes.scala [hunk 3]
```diff
@@ -16,55 +16,155 @@
...
 }
+
```
Comment (albertostratio) [inline src/main/scala/com/stratio/hermes/utils/Hermes.scala]: Could you please get rid of these blank lines

File: src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala [hunk 4]
```diff
@@ -24,10 +24,27 @@ import scala.util.Random
 object RandomHelper {
 
   /**
-   * Returns a random element from a list.
-   * @param l initial list
-   * @tparam T with the type of the list
-   * @return a random element of the list or None if the list is empty.
-   */
+    * Returns a random element from a list.
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala]: Indentation is wrong in comments. Configure your intellij. The same in the rest of files of your PR.

File: src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala [hunk 5]
```diff
@@ -24,10 +24,27 @@ import scala.util.Random
...
+    */
   def randomElementFromAList[T](l: List[T]): Option[T] = Random.shuffle(l).headOption
+
+  /**
+    * Returns length of a Integer element.
+    *
+    * @param n number to calculate length.
+    * @return size of the integer.
+    */
+  def numberOfDigitsFromANumber(n: Int): Int = if (n == 0) 1 else math.log10(math.abs(n)).toInt + 1
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala]: This method is not related with nothing of random and should not be here

File: src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala [hunk 6]
```diff
@@ -24,10 +24,27 @@ import scala.util.Random
...
+
+  /**
+    * Returns length of a Double element.
+    *
+    * @param n number to calculate length.
+    * @return size of the double.
+    */
+  def numberOfDigitsFromANumber(n: Double): Int = math.abs(n).toString.length - 1
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala]: This method is not related with nothing of random and should not be here

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 7]
```diff
@@ -25,54 +29,128 @@ class HermesTest extends FlatSpec with Matchers {
...
   it should "raise an exception when it tries to load a locale that don't exist" in {
     //scalastyle:off
-    val thrown = the [IllegalStateException] thrownBy Hermes("XY").Name.firstName()
+    val thrown = the[IllegalStateException] thrownBy Hermes("XY").Name.firstName()
+    //scalastyle:on
+    thrown.getMessage should equal(s"Error loading locale: /locales/name/XY.json")
+  }
+
+  it should "generate a random integer of 0 digit give it 0" in {
+
```
Comment (albertostratio) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala]: Could you please get rid of blank lines at the beggining and the end of all the tests?

File: src/main/scala/com/stratio/hermes/utils/Hermes.scala [hunk 8]
```diff
@@ -16,55 +16,155 @@
...
+      */
+    def middleName(): String = s"$firstName $firstName"
 
     /**
-     * Example: "Bruce".
-     * @return a first name.
-     */
-    def firstName() : String =
+      * Example: "Bruce".
+      *
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/utils/Hermes.scala]: delete this line

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 9]
```diff
@@ -25,54 +29,128 @@ class HermesTest extends FlatSpec with Matchers {
...
+    val hermesNum = Hermes("")
+    //scalastyle:off
+    forAll { (n: Int) =>
+      (n > 0 && n < 9) ==> (hermesNum.NumberGenerator.number(n, Positive) > 0)
+    }
+    forAll { (n: Int) =>
+      (n > 0 && n < 9) ==> (hermesNum.NumberGenerator.number(n, Negative) < 0)
+    }
+    val num: Int = hermesNum.NumberGenerator.number(2, Positive)
```
Comment (albertostratio) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala]: Does it make sense to check these unary conditions when you have check a lot of cases using scalacheck before? I'd just keep the forAll's in this test.

File: src/main/scala/com/stratio/hermes/utils/Hermes.scala [hunk 10]
```diff
@@ -16,55 +16,155 @@
...
+      Try(read[NameModel](getClass.getResourceAsStream(s"/locales/$unitName/$locale.json")))
         .getOrElse(throw new IllegalStateException(s"Error loading locale: /locales/$unitName/$locale.json"))
 
     /**
-     * Example: "Bruce Wayne".
-     * @return a full name.
-     */
-    def fullName() : String = s"$firstName $lastName"
+      * Example: "Bruce Wayne".
+      *
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/utils/Hermes.scala]: delete this line

File: src/main/scala/com/stratio/hermes/utils/Hermes.scala [hunk 11]
```diff
@@ -16,55 +16,155 @@
...
+      */
+    def fullName(): String = s"$firstName $lastName"
 
     /**
-     * Example: "Bruce Lex".
-     * @return a middle name.
-     */
-    def middleName() : String = s"$firstName $firstName"
+      * Example: "Bruce Lex".
+      *
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/utils/Hermes.scala]: delete this line

File: src/main/scala/com/stratio/hermes/utils/Hermes.scala [hunk 12]
```diff
@@ -16,55 +16,155 @@
...
+    def firstName(): String =
       RandomHelper.randomElementFromAList[String](nameModel.firstNames).getOrElse(throw new NoSuchElementException)
 
     /**
-     * Example: "Wayne".
-     * @return a last name.
-     */
-    def lastName() : String =
+      * Example: "Wayne".
+      *
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/utils/Hermes.scala]: delete this line

File: src/main/scala/com/stratio/hermes/utils/Hermes.scala [hunk 13]
```diff
@@ -16,55 +16,155 @@
...
   }
+
+  /**
+    * Generates random numbers.
+    */
+  object NumberGenerator {
+
+    /**
+      * Example: "number(3) 123".
+      *
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/utils/Hermes.scala]: delete this line

File: src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala [hunk 14]
```diff
@@ -24,10 +24,27 @@ import scala.util.Random
 object RandomHelper {
 
   /**
-   * Returns a random element from a list.
-   * @param l initial list
-   * @tparam T with the type of the list
-   * @return a random element of the list or None if the list is empty.
-   */
+    * Returns a random element from a list.
+    *
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala]: delete this line

File: src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala [hunk 15]
```diff
@@ -24,10 +24,27 @@ import scala.util.Random
...
+    * @param l initial list
+    * @tparam T with the type of the list
+    * @return a random element of the list or None if the list is empty.
+    */
   def randomElementFromAList[T](l: List[T]): Option[T] = Random.shuffle(l).headOption
+
+  /**
+    * Returns length of a Integer element.
+    *
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala]: delete this line

File: src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala [hunk 16]
```diff
@@ -24,10 +24,27 @@ import scala.util.Random
...
+    * @param n number to calculate length.
+    * @return size of the integer.
+    */
+  def numberOfDigitsFromANumber(n: Int): Int = if (n == 0) 1 else math.log10(math.abs(n)).toInt + 1
+
+  /**
+    * Returns length of a Double element.
+    *
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala]: delete this line

File: src/main/scala/com/stratio/hermes/utils/Hermes.scala [hunk 17]
```diff
@@ -16,55 +16,155 @@
...
+      * @return a last name.
+      */
+    def lastName(): String =
       RandomHelper.randomElementFromAList[String](nameModel.lastNames).getOrElse(throw new NoSuchElementException)
   }
+
+  /**
+    * Generates random numbers.
+    */
+  object NumberGenerator {
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/utils/Hermes.scala]: Change NumberGenerator by Number (the same as Name to follow an standard)

File: src/main/scala/com/stratio/hermes/utils/Hermes.scala [hunk 18]
```diff
@@ -16,55 +16,155 @@
...
+      * Example: "number(3) 123".
+      *
+      * @param n integer size.
+      * @return a random integer.
+      */
+    def number(n: Int): Int = {
+      assert(n>=0 && n<=9,throw new InvalidParameterException(s"$n must be between 0 and 9"))
+      if (n == 0) {
+        0
+      }
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/utils/Hermes.scala]: if {
} else {
}

You have

if {
}
else {
}

The same to all ifs of your PR

File: src/main/scala/com/stratio/hermes/utils/Sign.scala [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.stratio.hermes.utils
+
+sealed trait Sign
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/utils/Sign.scala]: This trait only wil be used in Hermes.scala. Move this trait to Hermes.scala and rename to NumberSign

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 20]
```diff
@@ -25,54 +29,128 @@ class HermesTest extends FlatSpec with Matchers {
...
+    hermesES.Name.nameModel.lastNames should contain(hermesES.Name.lastName)
   }
 
   it should "generate valid names: firstName lastName with EN and ES locales" in {
     val hermesEN = Hermes()
     val fullNameEN = hermesEN.Name.fullName
     fullNameEN should fullyMatch regex """[a-zA-Z]+ [a-zA-Z]+"""
-    hermesEN.Name.nameModel.firstNames should contain (fullNameEN.split(" ")(0))
-    hermesEN.Name.nameModel.lastNames should contain (fullNameEN.split(" ")(1))
+    hermesEN.Name.nameModel.firstNames should contain(fullNameEN.split(" ")(0))
```
Comment (anistal) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala]: Why are you deleting the spaces of contain?

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala (hunk 7 above)
Comment (anistal) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala]: delet this line

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 21]
```diff
@@ -25,54 +29,128 @@ class HermesTest extends FlatSpec with Matchers {
...
+    val thrown = the[IllegalStateException] thrownBy Hermes("XY").Name.firstName()
+    //scalastyle:on
+    thrown.getMessage should equal(s"Error loading locale: /locales/name/XY.json")
+  }
+
+  it should "generate a random integer of 0 digit give it 0" in {
+
+    val hermesNum = Hermes("")
+    hermesNum.NumberGenerator.number(0) shouldBe 0
+
```
Comment (anistal) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala:85]: delete this line

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 22]
```diff
@@ -25,54 +29,128 @@ class HermesTest extends FlatSpec with Matchers {
...
+  }
+
+  it should "generate a random integer when it passed the number of digit" in {
+
```
Comment (anistal) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala]: delete this line

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 23]
```diff
@@ -25,54 +29,128 @@ class HermesTest extends FlatSpec with Matchers {
...
+    val hermesNum = Hermes("")
+    forAll { (n: Int) =>
+      (n > 0 && n < 32) ==> (RandomHelper.numberOfDigitsFromANumber(hermesNum.NumberGenerator.number(n)) == n)
+    }
+
```
Comment (anistal) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala]: delete this line

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 24]
```diff
@@ -25,54 +29,128 @@ class HermesTest extends FlatSpec with Matchers {
...
+  }
+
+  it should "generate a random integer when it passed the number of digit and the sign" in {
+
```
Comment (anistal) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala]: delete this line

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 25]
```diff
@@ -25,54 +29,128 @@ class HermesTest extends FlatSpec with Matchers {
...
+      (n > 0 && n < 9) ==> (hermesNum.NumberGenerator.number(n, Negative) < 0)
+    }
+    val num: Int = hermesNum.NumberGenerator.number(2, Positive)
+    num > 0 shouldBe true
+    RandomHelper.numberOfDigitsFromANumber(num) shouldBe 2
+    val num2: Int = hermesNum.NumberGenerator.number(2, Negative)
+    RandomHelper.numberOfDigitsFromANumber(num2) shouldBe 2
+    num2 > 0 shouldBe false
+    //scalastyle:on
+
```
Comment (anistal) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala]: delete this line

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 26]
```diff
@@ -25,54 +29,128 @@ class HermesTest extends FlatSpec with Matchers {
...
+  }
+
+
+  it should "throw an InvalidParameterException when a negative digit is passed or greater than the VAL_MAX" in {
+
```
Comment (anistal) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala]: delete this line

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 27]
```diff
@@ -25,54 +29,128 @@ class HermesTest extends FlatSpec with Matchers {
...
+    val hermesNum = Hermes("")
+    //scalastyle:off
+    an[InvalidParameterException] should be thrownBy hermesNum.NumberGenerator.number(-2)
+    an[InvalidParameterException] should be thrownBy hermesNum.NumberGenerator.number(500)
+    an[InvalidParameterException] should be thrownBy hermesNum.NumberGenerator.decimal(-2)
+    an[InvalidParameterException] should be thrownBy hermesNum.NumberGenerator.decimal(2, -2)
+    an[InvalidParameterException] should be thrownBy hermesNum.NumberGenerator.decimal(2, 11)
+    //scalastyle:on
+
```
Comment (anistal) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala]: delete this line

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 28]
```diff
@@ -25,54 +29,128 @@ class HermesTest extends FlatSpec with Matchers {
...
+  }
+  it should "generate a random decimal of 0 digit give it 0.0" in {
+
```
Comment (anistal) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala]: delete this line

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 29]
```diff
@@ -25,54 +29,128 @@ class HermesTest extends FlatSpec with Matchers {
...
+    val hermesNum = Hermes("")
+    hermesNum.NumberGenerator.decimal(0) shouldBe 0.0
+    hermesNum.NumberGenerator.decimal(0, 0) shouldBe 0.0
+
```
Comment (anistal) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala]: delete this line

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 30]
```diff
@@ -25,54 +29,128 @@ class HermesTest extends FlatSpec with Matchers {
...
+  }
+
+  it should "generate a random decimal when it passed the number of digit" in {
+
```
Comment (anistal) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala]: delete this line

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 31]
```diff
@@ -25,54 +29,128 @@ class HermesTest extends FlatSpec with Matchers {
...
+    RandomHelper.numberOfDigitsFromANumber(hermesNum.NumberGenerator.decimal(2, 4)) shouldBe 6
+    RandomHelper.numberOfDigitsFromANumber(hermesNum.NumberGenerator.decimal(0, 2)) shouldBe 3
+    RandomHelper.numberOfDigitsFromANumber(hermesNum.NumberGenerator.decimal(2, 0)) shouldBe 3
+    RandomHelper.numberOfDigitsFromANumber(hermesNum.NumberGenerator.decimal(2, Positive)) shouldBe 4
+    RandomHelper.numberOfDigitsFromANumber(hermesNum.NumberGenerator.decimal(2, Negative)) shouldBe 4
+    RandomHelper.numberOfDigitsFromANumber(hermesNum.NumberGenerator.decimal(2, 1, Positive)) shouldBe 3
+    RandomHelper.numberOfDigitsFromANumber(hermesNum.NumberGenerator.decimal(2, 1, Negative)) shouldBe 3
     //scalastyle:on
-    thrown.getMessage should equal (s"Error loading locale: /locales/name/XY.json")
+
```
Comment (anistal) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala]: delete this line

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 32]
```diff
@@ -25,54 +29,128 @@ class HermesTest extends FlatSpec with Matchers {
...
   }
+
```
Comment (anistal) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala:137]: delete this line

File: src/main/scala/com/stratio/hermes/utils/Hermes.scala [hunk 33]
```diff
@@ -16,55 +16,155 @@
...
+      else {
+        val first = Random.nextInt(HermesConstants.ConstantDecimalValue - 1) + 1
+        val randSeq = first +: (1 until n).map { _ => Random.nextInt(HermesConstants.ConstantDecimalValue) }
+        BigInt(randSeq.mkString).toInt * randSign
+
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/utils/Hermes.scala]: delete this line

File: src/main/scala/com/stratio/hermes/utils/Hermes.scala [hunk 34]
```diff
@@ -16,55 +16,155 @@
...
+      }
+    }
+
+    /**
+      * Example: "number(3) 123".
+      *
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/utils/Hermes.scala]: delete this line

File: src/main/scala/com/stratio/hermes/utils/Hermes.scala [hunk 35]
```diff
@@ -16,55 +16,155 @@
...
+      } else {
+        val nonZero = Random.nextInt(HermesConstants.ConstantDecimalValue - 1) + 1
+        val randSeq = (1 until n).map { _ => Random.nextInt(HermesConstants.ConstantDecimalValue) } :+ nonZero
+        randSeq.mkString
+      }
+    }
+
+    /**
+      * Example: "number(3,Sign.-) -123".
+      *
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/utils/Hermes.scala]: delete this line

File: src/main/scala/com/stratio/hermes/utils/Hermes.scala [hunk 36]
```diff
@@ -16,55 +16,155 @@
...
+      * Example: "number(3) -> 123.456".
+      *
+      * @param n decimal part size.
+      * @return a random double with same integer and decimal part and random sign.
+      */
+    def decimal(n: Int): Double = (number(n).toString + "." + numberDec(n)).toDouble
+
+    /**
+      * Example: "decimal(3,Sign.-) -> -123.456".
+      *
```
Comment (anistal) [inline src/main/scala/com/stratio/hermes/utils/Hermes.scala]: delete this line

**Review by anistal — COMMENTED** [review]
The PR looks nice buy you have a lot of mistakes related with the coding style.

File: src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala (hunk 5 above)
Comment (albertostratio) [inline src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala]: This method is only used for testing purposes?? If so I'd not include it in this class

File: src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala (hunk 6 above)
Comment (albertostratio) [inline src/main/scala/com/stratio/hermes/helpers/RandomHelper.scala]: Same comment as above

**Review by albertostratio — CHANGES_REQUESTED** [review]

File: src/test/scala/com/stratio/hermes/utils/HermesTest.scala [hunk 37]
```diff
@@ -64,15 +67,94 @@ class HermesTest extends FlatSpec with Matchers {
...
+    val hermesNum = Hermes("")
+    //scalastyle:off
+    forAll { (n: Int) =>
+      hermesNum.Number.number(n, Positive) > 0
+    }
+    forAll { (n: Int) =>
+      hermesNum.Number.number(n, Negative) < 0
+    }
+    val num: Int = hermesNum.Number.number(2, Positive)
+    num > 0 shouldBe true
```
Comment (albertostratio) [inline src/test/scala/com/stratio/hermes/utils/HermesTest.scala]: Does it make sense to check these unary conditions when you have check a lot of cases using scalacheck before? I'd just keep the forAll's in this test.

**Review by albertostratio — CHANGES_REQUESTED** [review]

**Review by albertostratio — APPROVED** [review]
